sqlglot/transforms.py

```diff
@@ -398,6 +398,41 @@ def eliminate_full_outer_join(expression: exp.Expression) -> exp.Expression:
     return expression
 
 
+def move_ctes_to_top_level(expression: exp.Expression) -> exp.Expression:
+    """
+    Some dialects (e.g. Hive, T-SQL, Spark prior to version 3) only allow CTEs to be
+    defined at the top-level, so for example queries like:
+
+        SELECT * FROM (WITH t(c) AS (SELECT 1) SELECT * FROM t) AS subq
+
+    are invalid in those dialects. This transformation ensures all CTEs are moved
+    to the top level so that the final SQL code is valid from a syntax standpoint.
+
+    TODO: handle name clashes whilst moving CTEs (it can get quite tricky & costly).
+    """
+    # This transformation is only applied to the root expression, if it supports CTEs
+    if expression.parent or "with" not in expression.arg_types:
+        return expression
+
+    expression = expression.copy()
+    top_level_with = expression.args.get("with")
+    for node in expression.find_all(exp.With):
+        if node.parent is expression:
+            continue
+
+        inner_with = node.pop()
+        if not top_level_with:
+            top_level_with = inner_with
+            expression.set("with", top_level_with)
```

Comment:
what happens if the with alias collides?

Reply:
There's no name clash detection in this PR, so it'd shadow any name or if there were duplicate ctes it may crash depending on the engine I guess. This though was one of the core assumptions that allowed this to be simpler because checking for name clashes is more costly since cte names are global. To my understanding though having nested ctes with clashing names should be a rare case and we can always reiterate later?